### DISCUSSION TITLE
Use with Laravel Octane and Horizon

### DISCUSSION THREAD

**Discussion by robertpcontreras-ts** [spatie/once#79] [Q&A] [answered]
How does this function behave when run with Laravel Octane and Horizon?

The readme states that the cache only lasts for the duration of a request. Is this the same with Laravel octane? 

Also does the cache get reset for every new job processed within Horizon?

**Comment by freekmurze** [comment]
This function does not have any Laravel specific awareness or functionality. Under the hood is uses a `WeakMap` to store things. Learn more about that object and our internals here: https://www.youtube.com/watch?v=-lFyHJqzfFU&list=PLjzBMxW2XGTwEwWumYBaFHy1z4W32TcjU&index=13

You can flush the cache manually using

```
Spatie\Once\Cache::getInstance()->flush();
```

**Comment by stevebauman** [comment, marked as answer]
For anyone else looking into this -- I haven't implemented Octane yet with Spatie Once, but I believe we should be able to add a `OperationTerminated` listener to Laravel Octane's config to flush Spatie Once's cache:

https://github.com/laravel/octane/blob/bcd6ffe87bd389890fd7ef8ac24b46c579c8b434/config/octane.php#L103-L107

```php
// config/octane.php

OperationTerminated::class => [
    FlushTemporaryContainerInstances::class,
    // DisconnectFromDatabases::class,
    // CollectGarbage::class,

    FlushSpatieOnce::class, // <-- Here
],
```

```php
// app/Listeners/FlushSpatieOnce.php

use Spatie\Once\Cache;

class FlushSpatieOnce
{
    /**
     * Handle the event.
     *
     * @param  mixed  $event
     * @return void
     */
    public function handle($event): void
    {
        Cache::getInstance()->flush();
    }
}
```

The `OperationTerminated` event is a interface that is implemented by `TickTerminated`, `TaskTerminated` and `RequestTerminated`, which all get fired depending on the context of the application (receiving a request or executing a job/command).

**Reply by robertpcontreras-ts** [reply]
Thanks for this and definitely agree this would have the desired effect.

Any thoughts on how this would behave within Horizon workers (or any Laravel queue workers). I assume the once cache will persist during the life of the worker?

Also, am I right in assuming that if the above solution isn't implemented then the once cache will persist for the life of an octane worker?

This may be beneficial if you are happy for (or want) the once cache to persist for multiple requests for further performance improvements.

In this scenario the once cache will be reset after the task has executed the `--max-requests` value when starting the octane command.

**Reply by tm1000** [reply]
@robertpcontreras-ts you do the same thing but for Jobs (queue)

```php
//app/app/Providers/AppServiceProvider.php: boot()

Queue::looping(function () {
    \Spatie\Once\Cache::getInstance()->flush();
});
```

**Comment by davidjr82** [comment]
@stevebauman whenever I find a comment from you on an open source package, it helps to solve a problem, thank you for that.

@freekmurze as far as I have tested this solution (`FlushSpatieOnce` in `OperationTerminated` inside `config/octane.php`) it works fine. Should a message be written in the readme to point Octane users to this solution, saying that it is a "beta" solution for now until more people test it?

**Reply by freekmurze** [reply]
Feel free to send a PR to improve the readme.
